OpenRA.Mods.Common/Pathfinder/HierarchicalPathFinder.cs

```diff
@@ -718,24 +719,39 @@ static Grid GetGrid(CPos cellInGrid, Grid mapBounds)
 
 			pathFinderOverlay?.NewRecording(self, sources, target);
 
+			if (!world.Map.Contains(target))
+				return PathFinder.NoPath;
+
 			RebuildDirtyGrids();
 
 			var targetAbstractCell = AbstractCellForLocalCell(target);
 			if (targetAbstractCell == null)
 				return PathFinder.NoPath;
 
-			var sourcesWithReachableNodes = new List<CPos>(sources.Count);
-			var sourceEdges = new List<GraphEdge>(sources.Count);
+			// Unlike the target cell, the source cell is allowed to be an unreachable location.
+			// Instead, what matters is whether any cell adjacent to the source cell can be reached.
+			var sourcesWithReachableNodes = new List<(CPos Source, CPos AdjacentSource)>(sources.Count * CVec.Directions.Length);
+			var sourceEdges = new List<GraphEdge>(sources.Count * CVec.Directions.Length);
 			foreach (var source in sources)
 			{
-				var sourceAbstractCell = AbstractCellForLocalCell(source);
-				if (sourceAbstractCell == null)
```

Comment:
That's a great thought.

I've made this change - it also removes most of the performance loss - because most of the time the source is reachable and so we can avoid having to check all the adjacent locations. It also means we can restore the bidirectional search.

Have updated quite a few areas to make this happen and verified the test case is still working.